packages/hoprd/crates/hoprd-api/src/lib.rs

```diff
@@ -948,14 +955,25 @@ mod messages {
         #[serde_as(as = "Option<Vec<DisplayFromStr>>")]
         // #[validate(length(min=0, max=3))]        // TODO: issue in serde_as with validator -> no order is correct
         pub path: Option<Vec<PeerId>>,
-        #[validate(range(min=1, max=3))]
+        #[validate(range(min = 1, max = 3))]
         pub hops: Option<u16>,
     }
 
     #[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
     #[serde(rename_all = "camelCase")]
     struct SendMessageRes {
         pub challenge: HalfKeyChallenge,
+        pub timestamp: u128,
+    }
+
+    #[serde_as]
+    #[derive(Debug, Clone, serde::Serialize, serde::Deserialize, validator::Validate)]
+    #[serde(rename_all = "camelCase")]
+    struct GetMessageReq {
+        /// The message tag used to filter messages based on application
+        pub tag: u16,
+        /// Timestamp to filter messages received after this timestamp
+        pub timestamp: Option<u128>,
```

Comment:
Import `serde_with::DurationMilliSeconds` and then do:

```suggestion
        #[serde_as(as = "Option<DurationMilliSeconds<u64>>")]
        pub timestamp: Option<std::time::Duration>,
```

Also using `u128` is an overkill for timestamps with millisecond precision.